Implementation of a Command printing the release time of a build

DManstrator opened the pull request:
## Description

This PR adds a **command telling when a JDA build was released**.

If **no version** was provided, the **latest successful build** will be used instead.

**Potential Errors** like parsing the User Input and `IOException` of `JenkinsApi#getBuild` **are correctly handled**.

Used **Code Style was respected**.

## Example Preview
![Example Preview](http://puu.sh/DLe4z/39049cdfdd.png)

DManstrator committed:
Implementation of a Command printing the release time of a build

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -0,0 +1,124 @@
+            else if (ex instanceof NumberFormatException)
+                title = "Given input was not a valid build number!";
+            else
+                title = "Unknown Error occured!";
```
Logging this would be smart too (the full exception)

DManstrator replied:
Will do that. :+1:

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -0,0 +1,124 @@
+    @Override
+    public void dispatch(final User sender, final TextChannel channel, final Message message, final String content, final GuildMessageReceivedEvent event)
+    {
+        final EmbedBuilder eb = new EmbedBuilder();
```
This can be moved down to where its actually used to make it easier to find

DManstrator replied:
Well, it needs to be outside of the try-catch block (because it's in the catch as well). Should I create a second builder then?

kantenkugel replied:
Oh right. Didn't see that you used it in the catch block too

DManstrator replied:
Currently it looks like this:
```JAVA
final MessageEmbed failureEmbed = new EmbedBuilder().setAuthor("Error occured!", null, EmbedUtil.getJDAIconUrl())
    .setTitle(title, null)
    .setColor(Color.RED)
    .build();
reply(event, failureEmbed);
return;
```

```JAVA
// Return Info to User
final EmbedBuilder eb = new EmbedBuilder();
EmbedUtil.setColor(eb);

final MessageEmbed successEmbed = eb.setAuthor("Release Time of Version " + buildVersion, build.getUrl(), EmbedUtil.getJDAIconUrl())
    .setTitle(publishedTime, null)
    .build();

reply(event, successEmbed);
```

This seems better to me and as you said, easier to find. Normally I only use one embed if it is in the same scope but for this I can see why a second builder would be okay.

I guess since you marked this as resolved, you want it like it was originally with just one builder created in the beginning?

kantenkugel replied:
I marked as resolved since you did have a point to use it like you did. i don't the way you do it. Both ways are fine by me

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -0,0 +1,124 @@
+        final long buildTime = build.buildTime.toInstant().toEpochMilli();
+
+        CAL.setTimeInMillis(buildTime);
+        final String publishedTime = DFM.format(CAL.getTime());
```
Why do the complicated OffsetDateTime -> millis -> Calendar -> DateTimeFormatter instead of just directly formatting the OffsetDateTime?

DManstrator replied:
Didn't know it better tbh. lol
I will have a look onto it.

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -0,0 +1,124 @@
+public class DateVersionCommand extends Command
+{
+    private static final String DATE_FORMAT = "yyyy-MM-dd; HH:mm:ss";
+    private static final String[] ALIASES = { "date", "published" };
```
The command name is not needed in the aliases

DManstrator replied:
Yeah I thought so but I wasn't sure. Will be removed in the next commit.

kantenkugel reviewed (changes requested):
I think this is a good idea but why not just add that to the `changelog` command and add some aliases like `!build` (if it doesn't exist already). That way you have everything in one place.

DManstrator commented:
Didn't see your comment before, sry.

>I think this is a good idea but why not just add that to the changelog command and add some aliases like !build (if it doesn't exist already). That way you have everything in one place.

Don't know if you saw it but Alpaca kind of requested that, I just executed it ([Link to Discord Chat](https://discordapp.com/channels/125227483518861312/125227483518861312/593521965596278794)). :D

Actually I see this more like of a fun command where the changelog command is a serious one. If you really want I can do it like that but I don't think it's a good idea, the changelog already has a lot of contents and too many information aren't good imo.

DManstrator committed:
Implemented changes requested by the Review

Added Logging for Exception occurrences, simplified getting a formatted time based of a OffsetDateTime and removed command name from List of Aliases

DManstrator commented:
I wasn't home over the weekend and didn't want to commit without your response so I did it now, sorry for any inconveniences.

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -66,6 +65,9 @@ public void dispatch(final User sender, final TextChannel channel, final Message
         catch (IOException | NumberFormatException ex)
         {
+            final String fullStackTrace = getFullStackTrace(ex);
+            Bot.LOG.error("Exception in DateVersionCommand occured!" + System.lineSeparator() + fullStackTrace);
```
Logging errors in slf4j works differently: `Logger#error(String message, Throwable throwable)`. You don't need to evaluate the stacktrace nor concatenate anything yourself.

DManstrator replied:
Didn't know there was something for this so I made it myself. Actually nice to see that this function seems to do the exact same as my function did. :D

kantenkugel commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -103,6 +104,15 @@ else if (ex instanceof NumberFormatException)
+        builder.append(stacktrace);
+        return builder.toString();
+    }
+
```
Don't forget to remove this (`getFullStackTrace()`) once the logging is fixed

kantenkugel reviewed (changes requested):
Just a minor issue with using SLF4J incorrectly

DManstrator committed:
Correctly used the Logger class for logging an Exception

kantenkugel commented:
I'm just gonna let someone else (maybe Mighty) look over it before merging

DManstrator commented:
Yeah, that's fine. :)

Almighty-Alpaca commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/DateVersionCommand.java:
```diff
@@ -0,0 +1,121 @@
+
+public class DateVersionCommand extends Command
+{
+    private static final String DATE_FORMAT = "yyyy-MM-dd; HH:mm:ss";
```
Would be nice to include the timezone here as the time is kinda useless without it

kantenkugel replied:
That reminds me... I think i added the DurationUtils i wrote to Butler... Which means we could print the difference version-release -> now in some 1w, 3d, ... format as well (next to datetime).
Possibly needs to be expanded with month/year handling cuz i don't think i did those

DManstrator replied:
@Almighty-Alpaca Which Timezone should I use then? Or do you mean I should just add `O` to the [Pattern](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)?
@kantenkugel Can you explain the `includeIntermed` param of the `formatDuration` function? I saw no difference when trying out both.

Currently it looks like this:
![current state](http://puu.sh/DNrEl/4521795fde.png)
But this doesn't look very good to me. Any suggestions or are you fine with that?

>That said I also like @kantenkugel's idea to add it to the `changelog` command

Should I add it to the Changelog Command as well or should this be handled in another PR?

Edit:
>Possibly needs to be expanded with month/year handling cuz i don't think i did those

Yes, you would. Only milliseconds to days are in the `formatDuration` function (so even the week one is missing).

kantenkugel replied:
The `includeIntermed` parameter determines what happens to 0-values... with `includeIntermed` set to `true`, `10h 0m 0s` is possible. With `includeIntermed` set to `false`, it would just print `10h`. `0s` is only printed if no other value is >0

For this command, i would just include them... most of the time you won't have any 0-values anyway, so might as well go with all fields provided every time anyway.

For the changelog command, i would not add the duration btw. the datetime is enough there (timestamp of build)

For timezone, i would probably go with UTC.

Almighty-Alpaca replied:
Agree with Kant about UTC.

Looking at the last screenshot though I think it would be kinda cool to make it more text like, for example:

**Version 3.8.3_464 was released on 22/06/2019 at 18:08 (GMT)**
That was approximately 4 Days, 3 Hours and 32 Minutes ago.

DManstrator replied:
Okay, UTC is already used by now, `includeIntermed` is also set to `true`.

> Looking at the last screenshot though I think it would be kinda cool to make it more text like, for example:
> 
> **Version 3.8.3_464 was released on 22/06/2019 at 18:08 (GMT)**
> That was approximately 4 Days, 3 Hours and 32 Minutes ago.

Will do it like this for the DateVersionCommand. :+1: 
Where I would do it more generic in case you want to change this at some point.

My suggestion:
>**Release Time of Version <\<Version\>>:
><\<here some formatted time\>>**
>That was approximately <\<here some formatted duration\>> ago.

Is that okay for you? I just want it easy adjustable but if you say that this won't ever be changed, I will do it like you suggested it.

Other question: Here a function from the JenkinsChangelogProvider class:
```JAVA
    private static int extractBuild(String version)
    {
        int i = version.lastIndexOf('_');
        try
        {
            return Math.max(0, Integer.parseInt(version.substring(i + 1)));
        }
        catch(NumberFormatException ex)
        {
            return 0;
        }
    }
```
Would I be allowed to make it public? I could need that function for the getting the buildtime in the ChangelogCommand since the Changelog just provides the full version and sometimes it could be impossible to fetch a Build Number from that (higher failure percentage). Else I would make a copy of that function in the ChangelogCommand class.

### Current State
DateVersionCommand:
![DateVersionCommand](http://puu.sh/DNV23/b71740a008.png)

ChangelogCommand:
![ChangelogCommand](http://puu.sh/DNUVM/d3b4223bc9.png)

Almighty-Alpaca replied:
Sure, mine was only an idea. You don't have to do it exactly like that.

DManstrator replied:
> Sure, mine was only an idea. You don't have to do it exactly like that.

Well normally I expect that the devs have the last word on how they want the code to be and not the people requesting their code to be merged. :D

>Would I be allowed to make it public?

Any info on that? Then I can commit it real quick before going to bed and you can do further code checks. :D

Almighty-Alpaca replied:
I have never touched that piece of code but I don't see any reason why you could make it public

DManstrator replied:
Alright, then I will make a copy and don't make it public.

Almighty-Alpaca replied:
*couldn't, I meant I don't see any reason why you couldn't make it public, sry

DManstrator replied:
No worries, I already thought that, that's why I actually posted that comment. :D
And even if I would have used a copy, another commit and it's fixed so no worries. :)

Will be committed in the next 5 minutes.

Almighty-Alpaca reviewed (commented):
I suggested the command so you can show people how ancient their JDA versions are more easily. That said I also like @kantenkugel's idea to add it to the `changelog` command, although the focus there isn't on the date anymore. It can always be added to that command additionally.

Almighty-Alpaca commented:
Sure, mine was only an idea. You don't have to do it exactly like that.

DManstrator committed:
Added UTC TimeZone and Duration Comparison to DateVersionCommand and updated the Format of the printed DateTime

Also added the Build Time to the ChangelogCommand and refactored the code a bit

DManstrator commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/commands/commands/ChangelogCommand.java:
```diff
@@ -67,11 +77,33 @@ public void dispatch(final User sender, final TextChannel channel, final Message
+            {
+                final JenkinsBuild build = args == null
+                        ? JENKINS.getLastSuccessfulBuild()
+                        : JENKINS.getBuild(Integer.parseInt(args[0]));
```
I believe I made a tiny mistake here. Shouldn't `args[0]` not be the item name according to line 45?

I mean `VersionedItem#getChangelogProvider` always returns `null` at the moment so it will always stop executing with `"No Changelogs set up for " + item.getName()` (line 57) but if this is gonna be changed, this will "crash" (the `NumberFormatException` will be catched) when providing another Item since `args[0]` isn't a build number / an integer in that case.

I would fix that by replacing `args[0]` with `args[args.length - 1]`.

DManstrator replied:
@kantenkugel I talked already with Alpaca in the [JDA Server](https://discordapp.com/channels/125227483518861312/125227483518861312/609133986073280515) about this today, he has problems with the general code base right now so it would be great if you can also look over it (and maybe you can give me some response on [that one](#pullrequestreview-260090308) too). However, it "sounds plausible" for Alpaca and it's ready to be merged (after I updated the code based on my last review).

DManstrator commented on bot/src/main/java/com/almightyalpaca/discord/jdabutler/util/DateUtils.java:
```diff
@@ -0,0 +1,34 @@
+    public static final JenkinsApi JENKINS = JenkinsApi.JDA_JENKINS;
+    public static final DateTimeFormatter FORMATTER = getDateTimeFormatter();
+
+    public static DateTimeFormatter getDateTimeFormatter() {
```
The `{` should be in the next line to match the general code style.

DManstrator reviewed (commented):
https://github.com/Almighty-Alpaca/JDA-Butler/blob/6099bc71eb646e7de9ff5db2c05b02c295657d81/bot/src/main/java/com/almightyalpaca/discord/jdabutler/util/DateUtils.java#L17-L26

Maybe `DateTimeFormatter.ofPattern` should be called only once, doing the try-catch with a String (the pattern).

**Edit:**
Okay, don't know what I was thinking when commenting that. Adding ` (z)` twice looked kinda ugly so I thought this could be made better, also then only one `DateTimeFormatter.ofPattern` would be needed but since there is no method to check if a pattern is valid, the try-catch is required. Or do you know how this can be done better?
